app/forms/form_base.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class FormBase
+  include ActiveModel::Model
+  include ActiveModel::AttributeAssignment
+  include ActiveModel::Validations::Callbacks
+
+  attr_accessor :trainee, :user, :params, :store, :fields
+
+  delegate :id, :persisted?, to: :trainee
+
+  after_validation :track_validation_error, if: -> { user.present? && errors.any? }
+
+  def initialize(trainee, params: {}, user: nil, store: FormStore)
+    @user = user
+    @trainee = trainee
+    @params = params
+    @store = store
+    @fields = compute_fields
+    assign_attributes(fields)
+  end
+
+  def stash
+    valid? && store.set(id, form_store_key, fields)
+  end
+
+private
+
+  def compute_fields
+    raise NotImplementedError
+  end
+
+  def new_attributes
+    fields_from_store.merge(params).symbolize_keys
+  end
+
+  def track_validation_error
+    ValidationError.create!(form_object: self.class.name, user: user, details: validation_error_details.to_h)
```

Comment:
We should probably not raise here. If this fails for some reason we wouldn't want it return an error to the user.